File: intern/310.minimum-height-trees.js
/*
 * @lc app=leetcode id=310 lang=javascript
 *
 * [310] Minimum Height Trees
 */

// @lc code=start
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
const findMinHeightTrees = function (n, edges) {
    if (n === 1) {
        return [0];
    }
    /**
     * @type {Map<number, Set<number>>}
     */
    const connects = new Map();
    for (const [node1, node2] of edges) {
        const node1Edges = connects.get(node1) ?? new Set();
        node1Edges.add(node2);
        connects.set(node1, node1Edges);

        const node2Edges = connects.get(node2) ?? new Set();
        node2Edges.add(node1);
        connects.set(node2, node2Edges);
    }

    let eraseQueue = [];
    for (const [node1, node1Connects] of connects) {
        if (node1Connects.size === 1) {
            eraseQueue.push(node1);
        }
    }
    while (n > 2) {
        n -= eraseQueue.length;
        const eraseQueueCopy = [...eraseQueue];
        eraseQueue = [];
        for (const node of eraseQueueCopy) {
            const nodeConnects = connects.get(node);
            for (const nodeConnect of nodeConnects) {
                connects.get(nodeConnect).delete(node);
                if (connects.get(nodeConnect).size === 1) {
                    eraseQueue.push(nodeConnect);
                }
            }
            connects.delete(node);
        }
    }
    return eraseQueue;
};
// @lc code=end

File: Daily/310.minimum-height-trees.cpp
/*
 * @lc app=leetcode id=310 lang=cpp
 *
 * [310] Minimum Height Trees
 */
#include <map>
#include <unordered_set>
#include <vector>

using std::unordered_map;
using std::unordered_set;
using std::vector;

// @lc code=start
class Solution {
 public:
  vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
    unordered_map<int, unordered_set<int>> neighbors;
    for (const auto& edge : edges) {
      neighbors[edge[0]].insert(edge[1]);
      neighbors[edge[1]].insert(edge[0]);
    }

    unordered_set<int> nodes;
    for (int i = 0; i < n; i++) {
      nodes.insert(i);
    }

    vector<int> leaves;
    while (nodes.size() > 2) {
      for (const auto& node : nodes) {
        if (neighbors[node].size() == 1) {
          leaves.push_back(node);
        }
      }

      for (const auto& leaf : leaves) {
        for (const auto& neighbor : neighbors[leaf]) {
          neighbors[neighbor].erase(leaf);
        }
        neighbors.erase(leaf);
        nodes.erase(leaf);
      }

      leaves.clear();
    }

    return vector<int>(nodes.cbegin(), nodes.cend());
  }
};
// @lc code=end
